File: Valid Number.java
// case 1 : if we see [0-9] we reset the numberSeen flag.
// case 2 : we can only see '.' if we didn't see e or . .
// case 3 : we can only see 'e' if we didn't see 'e' but we see a number then we reset the numberSeen flag.
// case 4 : we can only see '+' and '-' in the beginning and after an 'e'
// case 5 : any other character break the validation.
class Solution {
public boolean isNumber(String s) {
    s = s.trim();

    boolean pointSeen = false;
    boolean eSeen = false;
    boolean numberSeen = false;

    for (int i = 0; i < s.length(); i++) {
        if(s.charAt(i) >= '0' && s.charAt(i) <= '9'){
            numberSeen = true;
        }
      else if(s.charAt(i) == '.'){
            if(eSeen || pointSeen){
                return false;
            }
            pointSeen = true;
        }
      else if(s.charAt(i) == 'e' || s.charAt(i) == 'E'){
            if(eSeen || !numberSeen){
                return false;
            }
            numberSeen = false;
            eSeen = true;
        }
      else if(s.charAt(i) == '-' || s.charAt(i) == '+'){
            if(i != 0 && s.charAt(i-1) != 'e'){
                return false;
            }
        }
      else {
            return false;
        }
    }
    return numberSeen; 
}
}

File: Valid_Number.cpp
class Solution {
public:
       bool isNumber(string s) {
    bool digit=false,e=false,dot=false;
    int sign=0;//for count  sign
    int n=s.size()-1;
    for(int i=0;i<=n;i++){
        char ch=s[i];
            // digit
        if(ch>='0'&&ch<='9')
                digit =true;
            // sign
        else if(ch=='+'||ch=='-'){
             if(sign>1)return false;
             else if(i>0&&s[i-1]!='e'&&s[i-1]!='E')return false;
             else if(i==n)return false;
                sign++;
                }
            //  dot
        else if(ch=='.'){
             if(e||dot)return false;
             else if(i==n&&!digit)return false;  
                dot=true;
            }
            //E or e
        else if(ch=='e'||ch=='E'){
             if(e||!digit||i==n)return false;
                e=true;
            }
        else{//if element is other than all i.e='a'
                return false;
            } 
        }    
    return true;
}
};
